megatron/core/datasets/gpt_dataset.py

```diff
@@ -669,23 +683,31 @@ def _get_ltor_masks_and_position_ids(
         position_ids = position_ids.clone()
 
     if reset_position_ids or reset_attention_mask:
-        # Find indices where EOD token is.
-        eod_index = position_ids[data == eod_token]
-        # Detach indices from positions if going to modify positions.
-        if reset_position_ids:
-            eod_index = eod_index.clone()
-
-        # Loop through EOD indices:
-        prev_index = 0
-        for j in range(eod_index.numel()):
-            i = eod_index[j]
-            # Mask attention loss.
-            if reset_attention_mask and attention_mask is not None:
-                attention_mask[0, (i + 1) :, : (i + 1)] = 0
-            # Reset positions.
-            if reset_position_ids:
-                position_ids[(i + 1) :] -= i + 1 - prev_index
-                prev_index = i + 1
+        # Find indices where BOS token is
+        bos_indices = (data == bos_token).nonzero(as_tuple=True)[0]
+
+        # Loop through BOS indices:
+        if bos_indices.numel() > 0:  # Only process if BOS tokens exist (codewise not strictly required but improves clarity.)
+            prev_index = 0
+            for i in bos_indices: 
+                if attention_mask is not None:
+                    # BOS token masking
+                    attention_mask[0, :, i] = 0  # No tokens can attend to BOS
+                    attention_mask[0, i, :] = 0  # BOS can't attend to any tokens
+                    
+                    if reset_attention_mask:
+                        # Block cross document attention
+                        attention_mask[0, i:, :i+1] = 0
+
+                        # If cross document attention is blocked, mask loss for last token of previous document
+                        if i > 0:  # Skip first BOS as there's no previous document
+                            loss_mask[i-1] = 0.0  # Mask loss for token right before BOS
+                    
+                # Reset positions.
+                if reset_position_ids:
+                    if i > 0:  # Skip first BOS as it's already at position 0
```

Comment:
1. if the formula is correct the additional if makes it more complicated but is not needed. you'll be basically doing - 0 (and it also works as a sanity check to make sure everything is working as expected). 
2. I am not fully familiar with the codebase but usually position ids start from 0. In that case, I think the formula should be (i - prev_index - 1) and prev_index should be set to -1 at the beginning. This doesn't matter much for RoPE though since it's relative but would be good to fix.